experimental/report/diagnostic.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package report
+
+import (
+	"fmt"
+
+	"github.com/bufbuild/protocompile/experimental/internal"
+)
+
+// Level represents the severity of a diagnostic message.
+type Level int8
+
+const (
+	// Internal compiler error. Indicates a panic within the compiler.
+	ICE Level = 1 + iota
+	// Red. Indicates a semantic constraint violation.
+	Error
+	// Yellow. Indicates something that probably should not be ignored.
+	Warning
+	// Cyan. This is the diagnostics version of "info".
+	Remark
+
+	noteLevel // Used internally within the diagnostic renderer.
+)
+
+// Tag is a diagnostic tag: a machine-readable identification for a diagnostic.
+//
+// Tags should be lowercase identifiers separated by dashes, e.g. my-error-tag.
+// If a package generates diagnostics with tags, it should expose those tags as
+// constants.
+type Tag string
+
+// Apply implements [DiagnosticOption].
+func (t Tag) Apply(d *Diagnostic) {
```

Comment:
Thankfully, your discoverability concerns aren't as big of a problem as you might fear.

For one, in the package docs, `Tag.Option` would have never sorted under `DiagnosticOption`. I do wish go had a way to spotlight implementations... but here we are (Rust has the same problem but the other way around, it shows too much!). The factory functions, thankfully, do not have this problem.

![Screenshot 2024-12-12 at 11 14 00 AM](https://github.com/user-attachments/assets/bf12c9c6-c9f7-4e8a-af75-9c76faf4ae66)

Also, autocomplete does find `Tag` constants in scope.

![Screenshot 2024-12-12 at 11 16 31 AM](https://github.com/user-attachments/assets/50f609c2-36d9-4a74-aa2e-b08b59b5bc43)

We can add

```go
func WithTag(t Tag) DiagnosticOption { return t }
```

if you want, but it feels redundant to me. The `With` prefix is also not consistent with the other option factories, and I'm not keen to add it...

Reply:
> Thankfully, your discoverability concerns aren't as big of a problem as you might fear.

Hmm. Those screenshots don't assuage my concerns at all. If anything, they make it more clear that `report.WithTag` would be better -- it would show up in the docs and it would show up to auto-complete after `report.` (which is the prefix that _all_ other options use). The auto-complete screenshot doesn't make sense: what are all of those other things in the suggestion list? How would I know that I can use the tag to produce an option? The inconsistency just makes it feel incoherent.

The only reason I suggested the `With` prefix is because `report.Tag` is already taken by the type. What about this: just use `string` as the type of a tag and then you can have `report.Tag(string) DiagnosticOption` as the way to set the tag. WDYT?

Reply:
I would *prefer* for Tag to be a type, for the sake of exported constants having a type that isn't string. But I don't think it matters too much in the end... if it turns out to be a problem we can revisit.